pkg/gitdir/gitdir.go

```diff
@@ -0,0 +1,437 @@
+package gitdir
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"sync"
+	"time"
+
+	git "github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	userName  = "Weave libgitops"
+	userEmail = "support@weave.works"
+
+	defaultBranch   = "master"
+	defaultRemote   = "origin"
+	defaultInterval = 30 * time.Second
+	defaultTimeout  = 1 * time.Minute
+)
+
+type GitDirectoryOptions struct {
+	// Options
+	Branch   string        // default "master"
+	Interval time.Duration // default 30s
+	Timeout  time.Duration // default 1m
+	// TODO: Support folder prefixes
+
+	// Authentication
+	// For HTTPS basic auth. The password should be e.g. a GitHub access token
+	Username, Password *string
+	// For Git SSH protocol. This is the bytes of e.g. ~/.ssh/id_rsa, given that ~/.ssh/id_rsa.pub is
+	// registered with and trusted by the Git provider. You can override the known hosts file using the
+	// SSH_KNOWN_HOSTS environment variable.
+	IdentityFileContent []byte
+}
+
+func (o *GitDirectoryOptions) Default() {
+	if o.Branch == "" {
+		o.Branch = defaultBranch
+	}
+	if o.Interval == 0 {
+		o.Interval = defaultInterval
+	}
+	if o.Timeout == 0 {
+		o.Timeout = defaultTimeout
+	}
+}
+
+// Create a new GitDirectory implementation
+func NewGitDirectory(url string, opts GitDirectoryOptions) (*GitDirectory, error) {
+	log.Info("Initializing the Git repo...")
+
+	// Default the options
+	opts.Default()
+
+	// Create a temporary directory for the clone
+	tmpDir, err := ioutil.TempDir("", "libgitops")
+	if err != nil {
+		return nil, err
+	}
+	log.Debugf("Created temporary git clone at %q", tmpDir)
```

Comment:
This might be misleading, the clone hasn't occurred at this point yet. Maybe `Created temporary directory for repository clone at %q` would be better?